### PR TITLE
Address short comings with printf for strings

### PR DESCRIPTION
**Pull request by Rekt3421**
This change fixes short comings encountered in 9 tests of the OpenCL-CTS as mentioned in issue #714 we still need to fix kernel descriptors.

### CODE REVIEW

**Comment by Rekt3421** [comment]
Next step is to add some unit tests

File: src/printf.cpp [hunk 1]
```diff
@@ -166,77 +166,121 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
...
-            // should be printed verbatim
-            printf_out << part_fmt;
-            break;
+    if (next_part > format_string.size()) {
+        // if there is no format string followed by a string to print just print
+        // the whole string itself.
+        if (descs.size() == 1) {
+            printf_out << format_string;
         }
+    } else {
```
Comment (rjodinchr) [inline src/printf.cpp:175]:
```suggestion
    if (next_part > format_string.size()) {
        // if there is no format just print the whole string itself.
        assert (descs.size() == 1);
        printf_out << format_string;
        printf("%s", printf_out.str().c_str());
        return;
    }
```

File: src/printf.cpp [hunk 2]
```diff
@@ -167,21 +167,54 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
...
+    }
     printf_out << format_string.substr(0, next_part);
 
     // Decompose the remaining format string into individual strings with
     // one format specifier each, handle each one individually
     size_t arg_idx = 0;
     while (next_part < format_string.size() - 1) {
-        // Get the part of the format string before the next format specifier
+        // Get the part of the format string before the next format
+        // specifier
```
Comment (rjodinchr) [inline src/printf.cpp:183]:
```suggestion
        // Get the part of the format string before the next format specifier
```

File: src/printf.cpp [hunk 3]
```diff
@@ -210,19 +243,24 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
...
             // Vector argument
             if (el_size == 0) {
-                // 'ele_size == 0' means that no length modifier has been used.
-                // According to the spec, this is an undefined behavior. Let's
-                // use the size coming from clspv and the vec_len to figure out
-                // the element size then.
+                // 'ele_size == 0' means that no length modifier has been
+                // used. According to the spec, this is an undefined
+                // behavior. Let's use the size coming from clspv and the
+                // vec_len to figure out the element size then.
```
Comment (rjodinchr) [inline src/printf.cpp:263]:
```suggestion
                // 'ele_size == 0' means that no length modifier has been used.
                // According to the spec, this is an undefined behavior. Let's
                // use the size coming from clspv and the vec_len to figure out
                // the element size then.
```

**Review by rjodinchr — APPROVED** [review]
Make sure you only modify what really needs to change.

File: tests/api/printf.cpp [hunk 4]
```diff
@@ -32,6 +32,20 @@ TEST_F(WithCommandQueueAndPrintf, SimplePrintf) {
     ASSERT_STREQ(m_printf_output.c_str(), message);
 }
 
+TEST_F(WithCommandQueueAndPrintf, SimpleFormatedPrintf) {
+    const char message[] = "";
+    char source[512];
+    sprintf(source, "kernel void test_printf() { printf(\"%%s\", \"\"); }");
```
Comment (rjodinchr) [inline tests/api/printf.cpp:38]: I don't get the point of using `sprintf` if they are no arguments after the format.

**Review by rjodinchr — APPROVED** [review]

**Comment by rjodinchr** [comment]
This is fixing CTS `printf string`, but `printf format_string` is regressing (segfault with `"\"%%\""`).

We should maybe try to address everything (`string` & `format_string`) together.

**Comment by Rekt3421** [comment]
current status 3 sub-subtests are still failing. One which is due to some issue with printf itself. 

I.e.  in strings subtest there is a subtest
```
printf(%%foo%%) 
```
is printing out 
```
%%foo%
```
which is a bit confusing since the two %% should be consumed as one. 

also in the same subtests for 
```
   printf("%s\n","foo\0foo");                                                           
``` 
the kernel descriptors are not properly populated for some reason.

a vector sub-subtest is also failing 
for which I do not have an explanation yet.

File: src/printf.cpp [hunk 5]
```diff
@@ -167,21 +184,66 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
...
+    }
     printf_out << format_string.substr(0, next_part);
 
     // Decompose the remaining format string into individual strings with
     // one format specifier each, handle each one individually
     size_t arg_idx = 0;
+    std::stack<char> quotes;
+    if (format_string.size() > 0) {
+        auto curr_quote = get_quote(format_string.substr(
+            0, next_part)); // checking if curr selection is inside a " or '.
```
Comment (Rekt3421) [inline src/printf.cpp:198]: Perhaps I should change this comment to "checking if curr selection has a " or '

File: src/printf.cpp [hunk 6]
```diff
@@ -13,12 +13,19 @@
...
+#include <stack>
 
 #include "printf.hpp"
 
 // Extract the conversion specifier from a format string
 char get_fmt_conversion(std::string_view fmt) {
     auto conversionSpecPos = fmt.find_first_of("diouxXfFeEgGaAcsp");
+    // Check if a valid conversion specifier was found
+    if (conversionSpecPos == std::string_view::npos) {
+        conversionSpecPos = -1;
```
Comment (rjodinchr) [inline src/printf.cpp:25]: Why setting `conversionSpecPos` to `-1`?

File: src/printf.cpp [hunk 7]
```diff
@@ -242,8 +318,8 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
 
     auto output = printf_out.str();
     if (printf_cb != nullptr) {
-        auto len = output.size();
-        printf_cb(output.c_str(), len, data >= data_end, printf_userdata);
+        printf_cb(output.c_str(), output.size(), data >= data_end,
+                  printf_userdata);
```
Comment (rjodinchr) [inline src/printf.cpp:322]: What does it fix?

Comment (Rekt3421) [reply src/printf.cpp:322]: nothing I just though it was more compact. Ill put it back to the way it was.

File: src/printf.cpp (hunk 6 above)
Comment (Rekt3421) [reply src/printf.cpp:25]: I was previously passing this as a ref to also get the position of the specifier, realized I don't need to, thanks for point this out;

**Review by rjodinchr — CHANGES_REQUESTED** [review]

File: src/printf.cpp [hunk 8]
```diff
@@ -165,14 +165,20 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
...
 
-    // Firstly print the part of the format string up to the first '%'
+    // Firstly print the part of the format string up to the first '%' if any
+    // otherwise print the whole string as is and move the data pointer to the
+    // end.
     size_t next_part = format_string.find_first_of('%');
-    printf_out << format_string.substr(0, next_part);
-
+    if (next_part == std::string::npos) {
+        printf_out << format_string;
```
[the remark this answers is not in the document]
Comment (Rekt3421) [reply src/printf.cpp:173]: this is what I get on my end if I do not have this change 

```
0)testing printf("","foo")
verifyOutputBuffer failed with kernel: 
__kernel void test9(void)
{
   printf("\n","foo");
}

 expected: 
 got:      
fo
```

File: src/printf.cpp [hunk 9]
```diff
@@ -210,8 +226,11 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
             // Special case for %s
             if (get_fmt_conversion(part_fmt) == 's') {
                 uint32_t string_id = read_buff<uint32_t>(data);
-                printf_out << print_part(
-                    part_fmt, descs.at(string_id).format_string.c_str(), size);
+                if (string_id < descs.size()) {
```
[the remark this answers is not in the document]
Comment (Rekt3421) [reply src/printf.cpp:229]: if I remove this, I get the following for string test_printf
```
1)testing printf("%s","")
terminate called after throwing an instance of 'std::out_of_range'
  what():  unordered_map::at
Aborted
```

[the remark this answers is not in the document]
Comment (Rekt3421) [reply src/printf.cpp]: okay I have removed it. Still seeing this failure but probably due to some issue on my end.

File: src/printf.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Rekt3421) [reply src/printf.cpp]: done. thanks

**Comment by Rekt3421** [comment]
@kpet to get this PR working correctly we need to update clspv with the fixes landed by David

File: src/printf.cpp [hunk 10]
```diff
@@ -181,7 +187,17 @@ void process_printf(char*& data, const printf_descriptor_map_t& descs,
         // Handle special cases
         if (part_end == part_start + 1) {
             printf_out << "%";
-            next_part = part_end + 1;
+            size_t next_format = format_string.find_first_of('%', part_end + 1);
```
Comment (rjodinchr) [inline src/printf.cpp:190]: Could you revert that part to what I propose?

**Review by rjodinchr — APPROVED** [review]

**Review by kpet — APPROVED** [review]

**Pull request merged**